lib/handshake_crypto.ml

```diff
@@ -24,56 +20,55 @@ let prf_mac = function
   | `ECDHE_RSA_WITH_AES_256_GCM_SHA384
   | `ECDHE_RSA_WITH_AES_256_CBC_SHA384
   | `ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
-  | `ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 -> (module SHA384 : S)
-  | _ -> (module SHA256 : S)
+  | `ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 -> (module Digestif.SHA384 : Digestif.S)
+  | _ -> (module Digestif.SHA256 : Digestif.S)
 
 let pseudo_random_function version cipher len secret label seed =
-  let labelled = Cstruct.of_string label <+> seed in
+  let labelled = label ^ seed in
   match version with
   | `TLS_1_1 | `TLS_1_0 ->
      let (s1, s2) = halve secret in
-     let md5 = p_hash (MD5.hmac, MD5.digest_size) s1 labelled len
-     and sha = p_hash (SHA1.hmac, SHA1.digest_size) s2 labelled len in
-     Mirage_crypto.Uncommon.Cs.xor md5 sha
+     let md5 = p_hash ((fun ~key s -> Digestif.MD5.(to_raw_string (hmac_string ~key s))), Digestif.MD5.digest_size) s1 labelled len
+     and sha = p_hash ((fun ~key s -> Digestif.SHA1.(to_raw_string (hmac_string ~key s))), Digestif.SHA1.digest_size) s2 labelled len in
+     Mirage_crypto.Uncommon.xor md5 sha
   | `TLS_1_2 ->
      let module D = (val (prf_mac cipher)) in
-     p_hash (D.hmac, D.digest_size) secret labelled len
+     p_hash ((fun ~key s -> D.(to_raw_string (hmac_string ~key s))), D.digest_size) secret labelled len
 
 let key_block version cipher len master_secret seed =
   pseudo_random_function version cipher len master_secret "key expansion" seed
 
 let hash version cipher data =
   match version with
-  | `TLS_1_0 | `TLS_1_1 -> MD5.digest data <+> SHA1.digest data
+  | `TLS_1_0 | `TLS_1_1 -> Digestif.(MD5.(to_raw_string (digest_string data)) ^ SHA1.(to_raw_string (digest_string data)))
   | `TLS_1_2 ->
-    let module H = (val (prf_mac cipher)) in
-    H.digest data
+    let module H = (val prf_mac cipher) in
+    H.(to_raw_string (digest_string data))
 
 let finished version cipher master_secret label ps =
-  let data = Cstruct.concat ps in
+  let data = String.concat "" ps in
   let seed = hash version cipher data in
   pseudo_random_function version cipher 12 master_secret label seed
 
 let divide_keyblock key mac iv buf =
-  let open Cstruct in
-  let c_mac, rt0 = split buf mac in
-  let s_mac, rt1 = split rt0 mac in
-  let c_key, rt2 = split rt1 key in
-  let s_key, rt3 = split rt2 key in
-  let c_iv , s_iv = split rt3 iv
+  let c_mac, rt0 = Core.split_str buf mac in
+  let s_mac, rt1 = Core.split_str rt0 mac in
+  let c_key, rt2 = Core.split_str rt1 key in
+  let s_key, rt3 = Core.split_str rt2 key in
+  let c_iv , s_iv = Core.split_str rt3 iv
```

Comment:
this is 10 string.sub. we can easily have fewer by using the offset and string.sub directly